crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -0,0 +1,808 @@
+        let chain_spec = Arc::new(
+            ChainSpecBuilder::default()
+                .chain(MAINNET.chain)
+                .genesis(MAINNET.genesis.clone())
+                .paris_activated()
+                .build(),
+        );
+        let (consensus_engine, env) = setup_consensus_engine(
+            chain_spec,
+            VecDeque::from([Err(StageError::ChannelClosed)]),
+            Vec::default(),
+        );
+        let mut rx = spawn_consensus_engine(consensus_engine);
+
+        // consensus engine is idle
+        std::thread::sleep(Duration::from_millis(100));
+        assert_matches!(rx.try_recv(), Err(TryRecvError::Empty));
+
+        // consensus engine is still idle
+        let _ = env.send_new_payload(SealedBlock::default().into());
+        assert_matches!(rx.try_recv(), Err(TryRecvError::Empty));
+
+        // consensus engine receives a forkchoice state and triggers the pipeline
+        let _ = env.send_forkchoice_updated(ForkchoiceState {
+            head_block_hash: H256::random(),
+            ..Default::default()
+        });
+        assert_matches!(
+            rx.await,
+            Ok(Err(BeaconEngineError::Pipeline(PipelineError::Stage(StageError::ChannelClosed))))
+        );
+    }
+
+    // Test that the consensus engine runs the pipeline again if the tree cannot be restored.
+    // The consensus engine will propagate the second result (error) only if it runs the pipeline
+    // for the second time.
+    #[tokio::test]
+    async fn runs_pipeline_again_if_tree_not_restored() {
+        let chain_spec = Arc::new(
+            ChainSpecBuilder::default()
+                .chain(MAINNET.chain)
+                .genesis(MAINNET.genesis.clone())
+                .paris_activated()
+                .build(),
+        );
+        let (consensus_engine, env) = setup_consensus_engine(
+            chain_spec,
+            VecDeque::from([
+                Ok(ExecOutput { stage_progress: 1, done: true }),
+                Err(StageError::ChannelClosed),
+            ]),
+            Vec::default(),
+        );
+        let rx = spawn_consensus_engine(consensus_engine);
+
+        let _ = env.send_forkchoice_updated(ForkchoiceState {
+            head_block_hash: H256::random(),
+            ..Default::default()
+        });
+        assert_matches!(
+            rx.await,
+            Ok(Err(BeaconEngineError::Pipeline(PipelineError::Stage(StageError::ChannelClosed))))
+        );
+    }
+
+    fn insert_blocks<'a, DB: Database>(db: &DB, mut blocks: impl Iterator<Item = &'a SealedBlock>) {
+        let mut transaction = Transaction::new(db).unwrap();
+        blocks
+            .try_for_each(|b| {
+                transaction
+                    .insert_block(SealedBlockWithSenders::new(b.clone(), Vec::default()).unwrap())
+            })
+            .expect("failed to insert");
+        transaction.commit().unwrap();
+    }
+
+    mod fork_choice_updated {
```

Comment:
I think this is missing a test where we get a meaningful payload? Whereas right now it's only testing for the bad payloads?

Reply:
technically that code path was checked in `unknown_head_hash`, but i added `valid_forkchoice` for clarity